File: Graph/detect_cycle_und_grp_dfs.cpp
class Solution {
    private:
    bool isCycleDFS(vector<vector<int>> &adj, vector<bool> &vis, int V, int n,int i, int parent){
        vis[i]=true;
        for(auto &it: adj[i]){ // can also write (auto it: adj[i]
            if(it==parent)
            continue;
            if(vis[it])
            return true;
            if(isCycleDFS(adj,vis,V,n,it,i))
            return true;
        }
        return false;
    }
  public:
    bool isCycle(int V, vector<vector<int>>& edges) {
        // Code here
        vector<vector<int>> adj(V);
        int n = edges.size();
        for(auto &edge:edges){ // can also write (auto edge : edges) but it will create copy, so extra memory usage
            int u = edge[0];
            int v = edge[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        vector<bool> vis(V,false);
         for(int i=0; i<V; i++){
             if(!vis[i] && isCycleDFS(adj,vis,V,n,i,-1))
             return true;
         }
         return false;
    }
};

File: Graph/detect_cycle_und_grp_dfs.java
import java.util.*;
public class detect_cycle_und_grp_dfs {
    
}
class Solution {
    public boolean isCycleDfs(List<List<Integer>> adj, boolean[] vis, int child, int parent){
        vis[child] = true;
        for(int it : adj.get(child)){
            if(it==parent)
            continue;
            if(vis[it])
            return true;
            if(isCycleDfs(adj, vis, it,child))
            return true;
        }
        return false;
    }
    public boolean isCycle(int V, int[][] edges) {
        List<List<Integer>> adj= new ArrayList<>();
        for(int i=0; i<V; i++){
            adj.add(new ArrayList<>());
        }
        for(int i=0; i<edges.length; i++){
            int u = edges[i][0];
            int v = edges[i][1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        boolean[] vis = new boolean[V];
        for(int i=0; i<V; i++){
            if(!vis[i] && isCycleDfs(adj,vis,i,-1))
            return true;
        }
        return false;
    }
}